### PR TITLE
Feature/COM-2599 Added Edit,delete,view,email and copy quote actions on quote 

### PR DESCRIPTION
**Pull request by devlucy**

### CODE REVIEW

File: scripts/modules/b2b-account/email-quote.js [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            templateName: "modules/b2b-account/quotes/email-quote",
+
+            initialize: function () {
+                Backbone.MozuView.prototype.initialize.apply(this, arguments);
+            },
+            render: function (accountId) {
+                var self = this;
+                self.model.set("accountId",accountId);
+                Backbone.MozuView.prototype.render.apply(this, arguments);
+                if(!accountInfo.isSalesRep)
```
Comment (sagarraut0007) [inline scripts/modules/b2b-account/email-quote.js]: add curly braces to if statements.

File: scripts/modules/b2b-account/email-quote.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+                $("#sendEmailText").val(accountInfo.email);
+                $(".modal-dialog").attr("id", "mz-email-modal");
+            $('[data-mz-action="handleSendemail"]').click(function() {
+               var emails; 
+               if(!self.model.attributes.inputEmails && !accountInfo.isSalesRep){
+               emails=$("#sendEmailText").val();
```
Comment (sagarraut0007) [inline scripts/modules/b2b-account/email-quote.js]: format your code properly. not seeing spaces in between the code. and entire formatting looks off to me.

File: scripts/modules/b2b-account/email-quote.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+               }
+               else{
+               emails=self.model.attributes.inputEmails;
+               }
+                var result = emails.split(",");
+                for(var i = 0;i < result.length && i < 20;i++){
```
Comment (sagarraut0007) [inline scripts/modules/b2b-account/email-quote.js]: Pull this max email number (20) out and save it into a variable on top. that way we can if we want to change this number later then we can do it easily.

File: scripts/modules/b2b-account/email-quote.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                }
+                if(invalidEmails.length===0){
+                var quote = new QuoteModels.Quote({
+                    id: self.model.get("accountId")
+                });
+                return quote.apiModel.email().then(function (res) {
+                   // to do after email sent successfully
+                    // emailAddress:validEmails
+                }, function (error) {
+                    if (error !== undefined){self.showMessageBar(error.message);}else{
```
Comment (sagarraut0007) [inline scripts/modules/b2b-account/email-quote.js]: format this properly. make it readable

File: scripts/modules/b2b-account/email-quote.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                        self.showMessageBar("Request Error");
+                    } 
+                }
+                );
+            }else{
+                for(var j = 0;j < invalidEmails.length;j++){
+                    self.showMessageBar(invalidEmails[j]);
```
Comment (sagarraut0007) [inline scripts/modules/b2b-account/email-quote.js]: If we have multiple invalid emails then we should not show multiple error messages. instead, we should show only one error message with all invalid emails in it.

File: scripts/modules/b2b-account/email-quote.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                }
+                if(invalidEmails.length ==  1){self.showMessageBar("Invalid Email Address ");} else{
```
Comment (sagarraut0007) [inline scripts/modules/b2b-account/email-quote.js]: Why we need this message again, we are already looped over invalidEmails[] and showed error messages. Why we need this if/else again. Show, only one error messages with all invalid emails in that message. with a proper readable message.

File: scripts/modules/b2b-account/email-quote.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                    self.showMessageBar("Invalid Email Addresses "+invalidEmails);
+                }
+            }
+            invalidEmails.length=0;
```
Comment (sagarraut0007) [inline scripts/modules/b2b-account/email-quote.js]: again formatting is wrong, looks right align. we don't do this generally.

File: scripts/modules/b2b-account/email-quote.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+              validEmails.length=0;
+                   emails.length=0;
+
+            });
+            $('[data-mz-action="checkEmail"]').on('keyup', function(e) {
+                $('.mz-messagebar').empty();
+              var tempVal= e.target.value;
```
Comment (sagarraut0007) [inline scripts/modules/b2b-account/email-quote.js]: give proper naming to a variable instead of tempVal

File: scripts/modules/b2b-account/email-quote.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+                if (tempVal.replace(/[^,]/g, "").length >= 20){
```
Comment (sagarraut0007) [inline scripts/modules/b2b-account/email-quote.js]: similar, use this 20 number from a variable from top. read my previous comment.

File: scripts/modules/b2b-account/quotes.js [hunk 10]
```diff
@@ -17,18 +17,18 @@ define([
     "modules/views-paging",
     'modules/editable-view',
     "modules/models-quotes",
-    "modules/b2b-account/account-address-search"], 
+    "modules/b2b-account/account-address-search",
+    "modules/b2b-account/email-quote"], 
     function ($, api, _, Hypr, Backbone, HyprLiveContext,
         CustomerModels, CartModels, B2BAccountModels, ProductModalViews,
         ProductPicker, ProductModels, WishlistModels, MozuGrid, MozuGridCollection,
```
Comment (sagarraut0007) [inline scripts/modules/b2b-account/quotes.js:26]: remove unnecessary models that we have injected here. Like i think 'ProductPicker' we haven't used in this file. but double-check and remove something which we haven't used here.

File: scripts/modules/b2b-account/quotes.js [hunk 11]
```diff
@@ -369,14 +381,60 @@ define([
...
-            this.set('copyQuoteView');
+        copyQuote: function (e, row) {
+            var self=this;
+            e.stopPropagation();
+            var quote = new QuoteModels.Quote({
+               id:row.id
+            });
+            return quote.apiModel.copy().then(function (res) {
+                if (isSalesRep) {
+                    window.location = '/selleraccount/quote/' + res.data.id + '/edit';
```
Comment (sagarraut0007) [inline scripts/modules/b2b-account/quotes.js]: save 'res.data.id' into a variable and then use it. name it properly.

File: scripts/modules/b2b-account/quotes.js [hunk 12]
```diff
@@ -369,14 +381,60 @@ define([
...
+        viewQuote: function (e, row) {
+            if (isSalesRep) {
+                window.location = '/selleraccount/quote/' + row.id;
+            }
+            else {
+                window.location = '/myaccount/quote/' + row.id;
+            }
+        },
+        showMessageBar: function (error) {
+            var self = this;
```
Comment (sagarraut0007) [inline scripts/modules/b2b-account/quotes.js:490]: i think we have to put showMessageBar() method in some common place and use it.

File: stylesheets/modules/b2b-account/email-quote.less [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+}
+.mz-text-margin
+{
+  margin-left: 2.5% !important;
+}
+
+.mz-button-footer {
+    text-align: end;
+}
+.mz-button-width
```
Comment (sagarraut0007) [inline stylesheets/modules/b2b-account/email-quote.less]: don't modify common css classes openly. wrap it around your parent class/container and then modify it. 
if not, then this will reflect in some other places as well where we have used these classes.

see following example-
.your-parent-container{
     .mz-button-width{
            //your styles here
      }
}

File: templates/modules/b2b-account/quotes/email-quote.hypr.live [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+      </div>
+{% block modal-body %}
+<div id="quotes-messages">
+    {% include "modules/common/message-bar" with model=model.messages %}
+</div>
+<div class="mz-l-stack-section mz-orderlist" style="clear:none">
+   
+    </br>
+        <div> 
+               <label class="mz-email-label">Email Addresses (Seperate Emails with Comma)</label>
```
Comment (sagarraut0007) [inline templates/modules/b2b-account/quotes/email-quote.hypr.live]: "Email Addresses (Seperate Emails with Comma)" add this to a labels files. 
In core theme, we have files 'en-US.json' and 'de-DE.json' (german version), add key/value there and use in templates like {{labels.key}}

File: templates/modules/mozuquotesgrid/table.hypr.live [hunk 15]
```diff
@@ -32,24 +32,75 @@
...
+                            {% if model.items[forloop.index0].customerAccountId === model.accountId and model.isAdmin == true %}
+                                 <button data-mz-action="toggleDropdown" class="dropbtn"></button>
                             <div id="quotesDropdown{{forloop.index}}" class="dropdown-content">
                             {% for action in model.rowActions %}
-                                {% if action.isHidden != true %}
-                                    <a href="#" data-mz-action="{{action.action}}">{{action.displayName}}</a>
-                                {% endif %}
-                            {% endfor %}                                
+                                      {% if action.action == "editQuote" %}
+                                       {% if row[5] =="InReview" or row[5] == "Expired" or row[5] == "Completed" %}{% else %}
```
Comment (sagarraut0007) [inline templates/modules/mozuquotesgrid/table.hypr.live]: These indexes might create problems in the future if we add /remove a column. think if we could do something about it.

**Review by sagarraut0007 — CHANGES_REQUESTED** [review]

File: scripts/modules/b2b-account/email-quote.js (hunk 1 above)
Comment (devlucy) [reply scripts/modules/b2b-account/email-quote.js]: done

File: scripts/modules/b2b-account/email-quote.js (hunk 2 above)
Comment (devlucy) [reply scripts/modules/b2b-account/email-quote.js]: done

File: scripts/modules/b2b-account/email-quote.js (hunk 3 above)
Comment (devlucy) [reply scripts/modules/b2b-account/email-quote.js]: done

File: scripts/modules/b2b-account/email-quote.js (hunk 4 above)
Comment (devlucy) [reply scripts/modules/b2b-account/email-quote.js]: done

File: scripts/modules/b2b-account/email-quote.js (hunk 5 above)
Comment (devlucy) [reply scripts/modules/b2b-account/email-quote.js]: done

File: scripts/modules/b2b-account/quotes.js (hunk 11 above)
Comment (devlucy) [reply scripts/modules/b2b-account/quotes.js]: done

File: scripts/modules/b2b-account/quotes.js (hunk 12 above)
Comment (devlucy) [reply scripts/modules/b2b-account/quotes.js:490]: Use case is different here. If I render self then it goes back to its default state and clears the input field.

File: stylesheets/modules/b2b-account/email-quote.less (hunk 13 above)
Comment (devlucy) [reply stylesheets/modules/b2b-account/email-quote.less]: done

File: templates/modules/b2b-account/quotes/email-quote.hypr.live (hunk 14 above)
Comment (devlucy) [reply templates/modules/b2b-account/quotes/email-quote.hypr.live]: done

File: templates/modules/mozuquotesgrid/table.hypr.live (hunk 15 above)
Comment (devlucy) [reply templates/modules/mozuquotesgrid/table.hypr.live]: done

File: scripts/modules/b2b-account/email-quote.js [hunk 16]
```diff
@@ -23,71 +23,77 @@ define([
...
+                        "id": self.model.get("accountId")
+                    };
 
-                }, function (error) {
-                    if (error.message !== undefined){
-                        self.showMessageBar(error);
-                    }else{
-                        self.showMessageBar({
-                            message: "Bad Request Error"
+                    var quoteId = self.model.get("accountId");
```
Comment (rakesh-dontula) [inline scripts/modules/b2b-account/email-quote.js]: why are you setting `accountId` as `quoteId`

**Review by rakesh-dontula — CHANGES_REQUESTED** [review]

File: scripts/modules/b2b-account/email-quote.js (hunk 16 above)
Comment (devlucy) [reply scripts/modules/b2b-account/email-quote.js]: Yes that's bad variable name I had used.I  have changed it to quoteId.

File: labels/de-DE.json [hunk 17]
```diff
@@ -307,8 +307,8 @@
   "mobileOrderConfirmationMessage": "{0} Bestellbenachrichtigungen. Wir werden Ihnen eine SMS senden, wenn sich Ihr Bestellstatus ändert. Die Meldungsfrequenz kann variieren. Antworte {1}, um dich abzumelden. Es können Nachrichten- und Datentarife gelten",
   "mobileOrderOptInMessage": "Vielen Dank, dass Sie sich für SMS-Benachrichtigungen für Ihre Curbside-Bestellung angemeldet haben. Wir werden Sie benachrichtigen, wenn Ihre Bestellung zur Abholung bereit ist.",
   "mobilePickLocation": "Ramasser / Expédier de:",
-  "mobilePickupReminder":"Erinnerung - Artikel in Ihrer Bestellung {0} können abgeholt werden. Überprüfen Sie die Bestellung unter {1}",
-  "mobileReadyForPickup":"Artikel in Ihrer Bestellung {0} können um {1} abgeholt werden. Überprüfen Sie die Bestellung unter {2}",
+  "mobilePickupReminder": "Erinnerung - Artikel in Ihrer Bestellung {0} können abgeholt werden. Überprüfen Sie die Bestellung unter {1}",
+  "mobileReadyForPickup": "Artikel in Ihrer Bestellung {0} können um {1} abgeholt werden. Überprüfen Sie die Bestellung unter {2}",
```
Comment (sagarraut0007) [inline labels/de-DE.json]: Your pr shouldn't have these changes. since you've not changed anything. Remove all of these changes from both labels files.

File: scripts/modules/b2b-account/delete-quote.js [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+                });
+            },
+            renderView: function () {
+                var self = this;
+                this.$el.html(this.template);
+                this.$el.modal({ show: true }); // dont show modal on instantiation
+            },
+            showMessageBar: function (meessage, isError) {
+                var self = this;
+                if (isError) { self.model.messagetype = "<ul class='is-showing mz-errors'>"; } else { self.model.messagetype = "<ul class='is-showing mz-success'>"; }
```
Comment (sagarraut0007) [inline scripts/modules/b2b-account/delete-quote.js]: make this if else statment formatted like below-
if(cond){
   //statement
} else{
   //statement
}

File: scripts/modules/b2b-account/email-quote.js [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+                            id: quoteId
+                        });
+                        return quote.apiModel.email(validEmails).then(function (res) {
+                            self.showMessageBar({ message: "Email has been sent successfully" }, false);
+                            setTimeout(function () {
+                                $('#closeModal').click();
+                            }, 4000);
+
+                        }, function (error) {
+                            if (error.message !== undefined) {
```
Comment (sagarraut0007) [inline scripts/modules/b2b-account/email-quote.js]: you dont have to check it against undefined just do a null check like below -

if(error.message){
          //statments here
}

File: scripts/modules/b2b-account/quotes.js [hunk 20]
```diff
@@ -335,16 +346,17 @@ define([
...
                 displayName: 'Status',
                 sortable: false
+               
             }
         ],
         rowActions: [
             {
-                displayName: 'Edit Quote',
-                action: 'editQuote'
+                displayName: 'View Quote',
```
Comment (sagarraut0007) [inline scripts/modules/b2b-account/quotes.js]: We also need to take this, displayName from label files. If you wanted to get values from labels files in javascript then you can use the following. Not sure this will work on not but give it a try. take the localized labels.

Hypr.getLabel('key')

File: scripts/modules/mozu-quotes-grid/mozuquotesgrid-view.js [hunk 21]
```diff
@@ -31,6 +37,23 @@ define(["modules/jquery-mozu",
                     };
                 });
             },
+            userHasBehavior: function (behaviorId) {
```
Comment (sagarraut0007) [inline scripts/modules/mozu-quotes-grid/mozuquotesgrid-view.js]: No need to write this comprehensive method, you can do shorthand them like below -

```
var behaviors = require.mozuData('user').behaviors;
this.model.set("isAdmin", behaviors.includes(1000) || behaviors.includes(1005));
```

**Review by sagarraut0007 — CHANGES_REQUESTED** [review]

File: scripts/modules/mozu-quotes-grid/mozuquotesgrid-view.js (hunk 21 above)
Comment (devlucy) [reply scripts/modules/mozu-quotes-grid/mozuquotesgrid-view.js]: done

File: scripts/modules/b2b-account/quotes.js (hunk 20 above)
Comment (devlucy) [reply scripts/modules/b2b-account/quotes.js]: Yes Its working. Done

File: scripts/modules/b2b-account/email-quote.js (hunk 19 above)
Comment (devlucy) [reply scripts/modules/b2b-account/email-quote.js]: done

File: scripts/modules/b2b-account/delete-quote.js (hunk 18 above)
Comment (devlucy) [reply scripts/modules/b2b-account/delete-quote.js]: done

File: labels/en-US.json [hunk 22]
```diff
@@ -379,8 +379,8 @@
   "mobileOrderConfirmationMessage": "{0} Order Alerts. We will text you when your order status changes. Msg freq may vary. Reply {1} to opt out. Msg&Data rates may apply",
   "mobileOrderOptInMessage": "Thank you for opting in to SMS alerts for your Curbside order. We will notify you when your order is ready for pickup.",
   "mobilePickLocation": "Pickup from/Ship from:",
-  "mobilePickupReminder":"Reminder - Items in your order {0} are ready for pickup. Review the order at {1}",
-  "mobileReadyForPickup":"Items in your order {0} are ready for pickup at {1}. Review the order at {2}",
+  "mobilePickupReminder": "Reminder - Items in your order {0} are ready for pickup. Review the order at {1}",
+  "mobileReadyForPickup": "Items in your order {0} are ready for pickup at {1}. Review the order at {2}",
```
Comment (ghost) [inline labels/en-US.json]: Don't make changes to labels you did not touch. This is core theme and our clients have to merge this. Revert all unnecessary whitespace changes in both label files.

File: labels/en-US.json [hunk 23]
```diff
@@ -754,5 +754,17 @@
   "width": "Width",
   "wrongCountryForType": "Please select a country supported for this contact type",
   "yourAccount": "Your Account",
-  "yourOrderNumber": "Your order number is <strong>{0}</strong>."
+  "yourOrderNumber": "Your order number is <strong>{0}</strong>.",
+  "emailLabel": "Email address(es) (separate emails with a comma)",
+  "emailQuote": "Email Quote",
+  "send": "Send",
+  "confirmDeleteQuote": "Are you sure want to delete {0}?",
```
Comment (ghost) [inline labels/en-US.json]: Your labels should be in alphabetical order not all added to the very end. You need to move these labels to the right spot in the file. Alphabetical order.

Fix both files.

File: scripts/modules/b2b-account/quotes.js [hunk 24]
```diff
@@ -251,6 +255,13 @@ define([
                 }
             }
         },
+        callbackForQuoteGridSelection:function(data){
```
Comment (ghost) [inline scripts/modules/b2b-account/quotes.js]: Ideally this should be called `selectRowInQuotesGrid` or `quotesGridRowSelected` as it better describes the action.

File: scripts/modules/mozu-quotes-grid/mozuquotesgrid-view.js (hunk 21 above)
Comment (ghost) [reply scripts/modules/mozu-quotes-grid/mozuquotesgrid-view.js]: If you're not using this method anymore then delete it. Don't say done and leave it here.

**Review by ghost — CHANGES_REQUESTED** [review]

File: scripts/modules/b2b-account/quotes.js (hunk 24 above)
Comment (devlucy) [reply scripts/modules/b2b-account/quotes.js]: I have made it `quotesGridRowSelected`.

File: labels/en-US.json (hunk 23 above)
Comment (devlucy) [reply labels/en-US.json]: I have ordered it alphabetically now.

File: labels/en-US.json (hunk 22 above)
Comment (devlucy) [reply labels/en-US.json]: done.

**Review by ghost — APPROVED** [review]
Some minor formatting issues but otherwise looks good now. Thank you.
